subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WindowsFileWatcherRegistryFactory.java

```diff
@@ -39,14 +45,22 @@ public WindowsFileWatcherRegistryFactory(WatchableFileSystemDetector watchableFi
     }
 
     @Override
-    protected FileWatcher createFileWatcher(BlockingQueue<FileWatchEvent> fileEvents) throws InterruptedException {
+    protected WindowsFileWatcher createFileWatcher(BlockingQueue<FileWatchEvent> fileEvents) throws InterruptedException {
         return fileEventFunctions.newWatcher(fileEvents)
             .withBufferSize(BUFFER_SIZE)
             .start();
     }
 
     @Override
-    protected FileWatcherUpdater createFileWatcherUpdater(FileWatcher watcher, WatchableHierarchies watchableHierarchies) {
-        return new HierarchicalFileWatcherUpdater(watcher, NO_VALIDATION, watchableHierarchies);
+    protected FileWatcherUpdater createFileWatcherUpdater(WindowsFileWatcher watcher, WatchableHierarchies watchableHierarchies) {
+        return new HierarchicalFileWatcherUpdater(watcher, NO_VALIDATION, watchableHierarchies, root -> invalidateMovedPaths(watcher, root));
+    }
+
+    private static SnapshotHierarchy invalidateMovedPaths(WindowsFileWatcher watcher, SnapshotHierarchy root) {
+        for (File movedPath : watcher.stopWatchingMovedPaths()) {
+            LOGGER.info("Dropping VFS state for moved path {}", movedPath.getAbsolutePath());
+            root = root.invalidate(movedPath.getAbsolutePath(), SnapshotHierarchy.NodeDiffListener.NOOP);
```

Comment:
Don't we need to update the watchers after invalidating parts of the VFS? I'd at least expect that we tell the watcher updater that we are not watching the moved path any more.

Reply:
I wanted to delete this comment in favor of this one on `HierarchicalFileWatcherUpdater`: https://github.com/gradle/gradle/pull/17922#discussion_r681741115